lib/pbench/server/database/models/metadata.py

```diff
@@ -0,0 +1,88 @@
+import enum
+import datetime
+from pbench.server.database.database import Database
+from sqlalchemy import Column, Integer, DateTime, ForeignKey, Text
+from sqlalchemy.orm import validates
+
+
+class MetadataKeys(enum.Enum):
+    FAVORITE = 1
+    SAVED = 2
+
+
+class Metadata(Database.Base):
+    """ Metadata Model for storing user metadata details """
+
+    # TODO: Think about the better name
+    __tablename__ = "metadata"
+
+    id = Column(Integer, primary_key=True, autoincrement=True)
+    created = Column(DateTime, nullable=False, default=datetime.datetime.now())
+    updated = Column(DateTime, nullable=False, default=datetime.datetime.now())
+    value = Column(Text, unique=False, nullable=False)
+    key = Column(Integer, nullable=False)
+    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=True)
+
+    @validates("key")
+    def evaluate_key(self, key, value):
+        return MetadataKeys[value].value
+
+    def get_json(self):
+        return {
+            "id": self.id,
+            "value": self.value,
+            "created": self.created,
+            "updated": self.updated,
+            "key": MetadataKeys(self.key).name,
+        }
+
+    @staticmethod
+    def get_protected():
+        return ["id", "created", "user_id"]
+
+    @staticmethod
+    def query(**kwargs):
+        query = Database.db_session.query(Metadata)
+        for attr, value in kwargs.items():
+            print(getattr(Metadata, attr), value)
+            query = query.filter(getattr(Metadata, attr) == value)
+        return query.all()
+
+    def add(self):
+        """
+        Add the current metadata object to the database
+        """
+        try:
+            Database.db_session.add(self)
+            Database.db_session.commit()
+        except Exception:
+            Database.db_session.rollback()
+            raise
+
+    def update(self, **kwargs):
+        """
+        Update the current user object with given keyword arguments
+        """
+        try:
+            for key, value in kwargs.items():
+                setattr(self, key, value)
+            Database.db_session.commit()
+        except Exception:
+            Database.db_session.rollback()
+            raise
+
+    @staticmethod
+    def delete(id):
+        """
+        Delete the metadata session with a given id
+        :param username:
```

Comment:
`:param id:` ... although we've developed a mix of docstring argument styles, and we really need to standardize. If we're going to use them, I'd prefer Google style. PEP actually would prefer relying on the signature with type hints. In any case, simply repeating the _names_ of parameters here doesn't really have any value. We should document what they _mean_ where it's not inherently obvious. (And I think we should start using type hints, but that's just me.)

Reply:
Fixed